README.md
This Jumpstart is a fork of [Jumpstart](https://github.com/excid3/jumpstart)

# Jumpstart Rails Template

All your Rails apps should start off with a bunch of great defaults. It's like Laravel Spark, for Rails.

Want to see how it works? Check out [the Jumpstart walkthrough video](https://www.youtube.com/watch?v=ssOZpISfIfI):

[![Jumpstart Ruby on Rails Template Walkthrough](https://i.imgur.com/pZDPbc7l.png)](https://www.youtube.com/watch?v=ssOZpISfIfI)

## Getting Started

Jumpstart is a Rails template, so you pass it in as an option when creating a new app.

#### Requirements

You'll need the following installed to run the template successfully:

* Ruby 2.5 or higher
* bundler - `gem install bundler`
* rails - `gem install rails`
* Database - we recommend Postgres, but you can use MySQL, SQLite3, etc
* Redis - For ActionCable support
* ImageMagick or libvips for ActiveStorage variants
* Yarn - `brew install yarn` or [Install Yarn](https://yarnpkg.com/en/docs/install)
* Foreman (optional) - `gem install foreman` - helps run all your processes in development

#### Creating a new app

```bash
rails new myapp -d postgresql -m https://raw.githubusercontent.com/przbadu/jumpstart/master/template.rb
```

Or if you have downloaded this repo, you can reference template.rb locally:

```bash
rails new myapp -d postgresql -m template.rb
```

❓Having trouble? Try adding `DISABLE_SPRING=1` before `rails new`. Spring will get confused if you create an app with the same name twice.

#### Running your app

```bash
bin/dev
```

You can also run them in separate terminals manually if you prefer.

A separate `Procfile` is generated for deploying to production on Heroku.

#### Authenticate with social networks

We use the encrypted Rails Credentials for app_id and app_secrets when it comes to omniauth authentication. Edit them as so:

```
EDITOR=vim rails credentials:edit
```

Make sure your file follow this structure:

```yml
secret_key_base: [your-key]
development:
  github:
    app_id: something
    app_secret: something
    options:
      scope: 'user:email'
      whatever: true
production:
  github:
    app_id: something
    app_secret: something
    options:
      scope: 'user:email'
      whatever: true
```

With the environment, the service and the app_id/app_secret. If this is done correctly, you should see login links
for the services you have added to the encrypted credentials using `EDITOR=vim rails credentials:edit`

#### Enabling Admin Panel
App uses `madmin` [gem](https://github.com/excid3/madmin), so you need to run the madmin generator:

```
rails g madmin:install
```

This will install Madmin and generate resources for each of the models it finds.
#### Redis set up
##### On OSX
```
brew update
brew install redis
brew services start redis
```
##### Ubuntu
```
sudo apt-get install redis-server
```

#### Cleaning up

```bash
rails db:drop
spring stop
cd ..
rm -rf myapp
```


CHANGELOG.md
### 2023-07-19

* Support for [Tailwind css](https://tailwindcss.com) with [DaisyUI](https://daisyui.com)
* Updated all devise templates to use tailwind css

### 2021-12-27

* Support Rails 7.0
* Drop support for Rails 5.2 and earlier

### 2021-05-13

* Upgrade to Bootstrap 5
* Remove data-confirm-modal, not yet bootstrap 5 compatible
* Drop jQuery
* Include devise views directly

### 2021-03-04

* Switch to Madmin for admin area

### 2020-10-24

* Rescue from git configuration exception

### 2020-08-20

* Add tests for generating postgres and mysql apps

### 2020-08-07

* Refactor notifications to use the [Noticed gem](https://github.com/excid3/noticed)

### 2019-02-28

* Adds support for Rails 6.0
* Move all Javascript to Webpacker for Rails 5.2 and 6.0
  * Use Bootstrap, data-confirm-modal, and local-time from NPM packages
  * ProvidePlugin sets jQuery, $, and Rails variables for webpacker
* Use https://github.com/excid3/administrate fork of Administrate
  * Adds fix for zeitwerk autoloader in Rails 6
  * Adds support for virtual attributes
* Add Procfile, Procfile.dev and .foreman configs
* Add welcome message and instructions after completion

### 2019-01-02 and before

* Original version of Jumpstart
* Supported Rails 5.2 only


test/template_test.rb
require 'minitest/autorun'

class TemplateTest < Minitest::Test
  def setup
    system('[ -d test_app ] && rm -rf test_app')
  end

  def teardown
    setup
  end

  def test_generator_succeeds
    output, _err = capture_subprocess_io do
      system('DISABLE_SPRING=1 SKIP_GIT=1 rails new test_app -m template.rb')
    end
    assert_includes output, 'Jumpstart app successfully created!'
  end

  # TODO: Fix these tests on CI so they don't fail on db:create
  #
  # def test_generator_with_postgres_succeeds
  #   output, err = capture_subprocess_io do
  #     system("DISABLE_SPRING=1 SKIP_GIT=1 rails new test_app -m template.rb -d postgresql")
  #   end
  #   assert_includes output, "Jumpstart app successfully created!"
  # end

  # def test_generator_with_mysql_succeeds
  #   output, err = capture_subprocess_io do
  #     system("DISABLE_SPRING=1 SKIP_GIT=1 rails new test_app -m template.rb -d mysql")
  #   end
  #   assert_includes output, "Jumpstart app successfully created!"
  # end
end


template.rb
require 'fileutils'
require 'shellwords'

# Copied from: https://github.com/mattbrictson/rails-template
# Add this template directory to source_paths so that Thor actions like
# copy_file and template resolve against our source files. If this file was
# invoked remotely via HTTP, that means the files are not present locally.
# In that case, use `git clone` to download them to a local temporary dir.
def add_template_repository_to_source_path
  if __FILE__ =~ %r{\Ahttps?://}
    require 'tmpdir'
    source_paths.unshift(tempdir = Dir.mktmpdir('jumpstart-'))
    at_exit { FileUtils.remove_entry(tempdir) }
    git clone: [
      '--quiet',
      'https://github.com/przbadu/jumpstart.git',
      tempdir
    ].map(&:shellescape).join(' ')

    if (branch = __FILE__[%r{jumpstart/(.+)/template.rb}, 1])
      Dir.chdir(tempdir) { git checkout: branch }
    end
  else
    source_paths.unshift(File.dirname(__FILE__))
  end
end

def rails_version
  @rails_version ||= Gem::Version.new(Rails::VERSION::STRING)
end

def rails_6_or_newer?
  Gem::Requirement.new('>= 6.0.0.alpha').satisfied_by? rails_version
end

def add_gems
  add_gem 'cssbundling-rails'
  add_gem 'tailwindcss-rails'
  add_gem 'devise', '~> 4.9'
  add_gem 'friendly_id', '~> 5.4'
  add_gem 'jsbundling-rails'
  add_gem 'madmin'
  add_gem 'name_of_person', '~> 1.1'
  add_gem 'noticed', '~> 1.4'
  add_gem 'omniauth-facebook', '~> 8.0'
  add_gem 'omniauth-github', '~> 2.0'
  add_gem 'omniauth-twitter', '~> 1.4'
  add_gem 'pretender', '~> 0.3.4'
  add_gem 'pundit', '~> 2.1'
  add_gem 'sidekiq', '~> 6.2'
  add_gem 'sitemap_generator', '~> 6.1'
  add_gem 'whenever', require: false
  add_gem 'responders', github: 'heartcombo/responders', branch: 'main'
end

def set_application_name
  # Add Application Name to Config
  environment 'config.application_name = Rails.application.class.module_parent_name'

  # Announce the user where they can change the application name in the future.
  puts 'You can change application name inside: ./config/application.rb'
end

def add_users
  route "root to: 'home#index'"
  generate 'devise:install'

  environment "config.action_mailer.default_url_options = { host: 'localhost', port: 3000 }", env: 'development'
  generate :devise, 'User', 'first_name', 'last_name', 'announcements_last_read_at:datetime', 'admin:boolean'

  # Set admin default to false
  in_root do
    migration = Dir.glob('db/migrate/*').max_by { |f| File.mtime(f) }
    gsub_file migration, /:admin/, ':admin, default: false'
  end

  gsub_file 'config/initializers/devise.rb', /  # config.secret_key = .+/, '  config.secret_key = Rails.application.credentials.secret_key_base' if Gem::Requirement.new('> 5.2').satisfied_by? rails_version

  inject_into_file('app/models/user.rb', 'omniauthable, :', after: 'devise :')
end

def add_authorization
  generate 'pundit:install'
end

def default_to_esbuild
  return if options[:javascript] == 'esbuild'

  return if options[:skip_javascript]

  @options = options.merge(javascript: 'esbuild')
end

def add_javascript
  run 'yarn add local-time esbuild-rails trix @hotwired/stimulus @hotwired/turbo-rails @rails/activestorage @rails/ujs @rails/request.js chokidar'
  run 'yarn add -D daisyui@latest @tailwindcss/forms @tailwindcss/aspect-ratio @tailwindcss/typography @tailwindcss/container-queries'
end

def copy_templates
  remove_file 'app/assets/stylesheets/application.css'
  remove_file 'app/javascript/application.js'
  remove_file 'app/javascript/controllers/index.js'
  remove_file 'Procfile.dev'
  remove_file 'config/tailwind.config.js'

  copy_file 'Procfile'
  copy_file 'Procfile.dev'
  copy_file '.foreman'
  copy_file 'esbuild.config.mjs'
  copy_file 'app/javascript/application.js'
  copy_file 'app/javascript/controllers/index.js'
  copy_file 'config/tailwind.config.js'

  directory 'app', force: true
  directory 'config', force: true
  directory 'lib', force: true

  route "get '/terms', to: 'home#terms'"
  route "get '/privacy', to: 'home#privacy'"
end

def add_sidekiq
  environment 'config.active_job.queue_adapter = :sidekiq'

  insert_into_file 'config/routes.rb',
                   "require 'sidekiq/web'\n\n",
                   before: 'Rails.application.routes.draw do'

  content = <<~RUBY
    authenticate :user, lambda { |u| u.admin? } do
      mount Sidekiq::Web => '/sidekiq'

      namespace :madmin do
        resources :impersonates do
          post :impersonate, on: :member
          post :stop_impersonating, on: :collection
        end
      end
    end
  RUBY
  insert_into_file 'config/routes.rb', "#{content}\n", after: "Rails.application.routes.draw do\n"
end

def add_announcements
  generate 'model Announcement published_at:datetime announcement_type name description:text'
  route 'resources :announcements, only: [:index]'
end

def add_notifications
  route 'resources :notifications, only: [:index]'
end

def add_multiple_authentication
  insert_into_file 'config/routes.rb', ', controllers: { omniauth_callbacks: "users/omniauth_callbacks" }', after: '  devise_for :users'

  generate 'model Service user:references provider uid access_token access_token_secret refresh_token expires_at:datetime auth:text'

  template = ''"
  env_creds = Rails.application.credentials[Rails.env.to_sym] || {}
  %i{ facebook twitter github }.each do |provider|
    if options = env_creds[provider]
      config.omniauth provider, options[:app_id], options[:app_secret], options.fetch(:options, {})
    end
  end
  "''.strip

  insert_into_file 'config/initializers/devise.rb', '  ' + template + "\n\n", before: '  # ==> Warden configuration'
end

def add_whenever
  run 'wheneverize .'
end

def add_friendly_id
  generate 'friendly_id'
  insert_into_file(Dir['db/migrate/**/*friendly_id_slugs.rb'].first, '[5.2]', after: 'ActiveRecord::Migration')
end

def add_sitemap
  rails_command 'sitemap:install'
end

def add_tailwindcss
  rails_command 'tailwindcss:install'
end

def add_announcements_css
  insert_into_file 'app/assets/stylesheets/application.tailwind.css', '@import "jumpstart/announcements";'
end

def add_esbuild_script
  build_script = 'node esbuild.config.mjs'

  case `npx -v`.to_f
  when 7.1...8.0
    run %(npm set-script build "#{build_script}")
    run %(yarn build)
  when (8.0..)
    run %(npm pkg set scripts.build="#{build_script}")
    run %(yarn build)
  else
    say %(Add "scripts": { "build": "#{build_script}" } to your package.json), :green
  end
end

def add_gem(name, *options)
  gem(name, *options) unless gem_exists?(name)
end

def gem_exists?(name)
  IO.read('Gemfile') =~ /^\s*gem ['"]#{name}['"]/
end

puts 'Please use Rails 6.0 or newer to create a Jumpstart application' unless rails_6_or_newer?

# Main setup
add_template_repository_to_source_path
default_to_esbuild
add_gems

after_bundle do
  set_application_name
  add_users
  add_authorization
  add_javascript
  add_announcements
  add_notifications
  add_multiple_authentication
  add_sidekiq
  add_friendly_id
  add_tailwindcss
  add_whenever
  add_sitemap
  add_announcements_css
  rails_command 'active_storage:install'

  # Make sure Linux is in the Gemfile.lock for deploying
  run 'bundle lock --add-platform x86_64-linux'

  copy_templates

  add_esbuild_script

  # Commit everything to git
  unless ENV['SKIP_GIT']
    git :init
    git add: '.'
    # git commit will fail if user.email is not configured
    begin
      git commit: %( -m 'Initial commit' )
    rescue StandardError => e
      puts e.message
    end
  end

  say
  say 'Jumpstart app successfully created!', :blue
  say
  say 'To get started with your new app:', :green
  say "  cd #{original_app_name}"
  say
  say '  # Update config/database.yml with your database credentials'
  say
  say '  rails db:create'
  say '  rails g noticed:model'
  say '  rails db:migrate'
  say '  rails g madmin:install # Generate admin dashboards'
  say '  gem install foreman'
  say '  bin/dev'
end
